review/database.py

```diff
@@ -0,0 +1,541 @@
+    return filter
+
+
+Teacher.add_used_filter_generator(teacher_teacherreview_filter_generator)
+Teacher.add_used_filter_generator(teacher_subjectreview_filter_generator)
+
+
+class SubjectRelevance(database.base):
+    """Holds user votes for subject reviews.
+
+    Args:
+        voter_id: User's Discord ID
+        vote: True if positive, False if negative
+        review: SubjectReview IDX
+    """
+
+    __tablename__ = "school_review_subject_relevance"
+
+    voter_id = Column(BigInteger, primary_key=True)
+    vote = Column(Boolean, default=False)
+    review = Column(
+        Integer,
+        ForeignKey("school_review_subject_review.idx", ondelete="CASCADE"),
+        primary_key=True,
+    )
+
+    @staticmethod
+    def reset(review_id: int):
+        result = session.query(SubjectRelevance).filter_by(review=review_id).delete()
+        session.commit()
+        return result
+
+    def __repr__(self) -> str:
+        return (
+            f'<{self.__class__.__name__} voter_id="{self.voter_id}" vote="{self.vote}" '
+            f'review="{self.review}">'
+        )
+
+    def dump(self) -> dict:
+        return {
+            "voter_id": self.voter_id,
+            "vote": self.vote,
+            "review": self.review,
+        }
+
+
+class SubjectReview(database.base):
+    """Holds information about subject reviews and all the logic.
+
+    Args:
+        idx: Unique review ID
+        guild_id: Guild ID
+        author_id: Author's Discord ID
+        anonym: Show or hide author name
+        subject_id: IDX of Subject
+        grade: Grade as in school (should be 1-5)
+        text_review: Text of review
+        created: Date of creation
+        updated: Date of update
+        subject: Relationship with Subject
+        guarantor: Relationship with Teacher (subject guarantor)
+        relevance: Relationship with Votes (SubjectRelevance)
+        upvotes: Count of positive votes
+        downvotes: Count of negative votes
+    """
+
+    __tablename__ = "school_review_subject_review"
+    __table_args__ = (
+        UniqueConstraint(
+            "author_id",
+            "guild_id",
+            "subject_id",
+            name="guild_id_author_id_subject_id_unique",
+        ),
+    )
+
+    idx = Column(Integer, primary_key=True, autoincrement=True)
+    guild_id = Column(BigInteger)
+    author_id = Column(BigInteger)
+    anonym = Column(Boolean, default=True)
+    subject_id = Column(
+        Integer, ForeignKey("school_dataset_subject.idx", ondelete="CASCADE")
+    )
+    guarantor_id = Column(
+        Integer, ForeignKey("school_dataset_teacher.idx", ondelete="SET NULL")
+    )
+    grade = Column(Integer, default=0)
+    text_review = Column(String, default=None)
+    created = Column(Date)
+    updated = Column(Date)
+    subject = relationship("Subject")
+    guarantor = relationship(lambda: Teacher)
+    relevance = relationship("SubjectRelevance", cascade="all, delete-orphan, delete")
+
```

Comment:
Or... To reduce complexity just ask the user to provide that information un the `text_review`

Reply:
The guarantor is there because if the guarantor changes, the subject can dramatically change. However for each individual teacher there are teacher reviews and they can be mentioned in the text review.

Reply:
I'm just talking about the teachers, not tha guarantor. Guarantor should stay but people with different teachers will also have different experiences so I'm thinking about ways to include them too

Reply:
The teachers have their own review and can be reviewed separately in the text.